### PR TITLE
[SPARK-40910][SQL] Replace UnsupportedOperationException with SparkUnsupportedOperationException

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr aims to replace UnsupportedOperationException with SparkUnsupportedOperationException.

### Why are the changes needed?
1.When I work on https://issues.apache.org/jira/browse/SPARK-40889,
I found `QueryExecutionErrors.unsupportedPartitionTransformError` throw **UnsupportedOperationException**(but not **SparkUnsupportedOperationException**), it seem not to fit into the new error framework.
https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala#L71-L72

2.`QueryExecutionErrors.unsupportedPartitionTransformError` throw SparkUnsupportedOperationException, but UT catch `UnsupportedOperationException`.
https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala#L288-L301

https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala#L904-L909

https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/core/src/main/scala/org/apache/spark/SparkException.scala#L144-L154

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existed UT.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 1]
```diff
@@ -292,13 +293,12 @@ class PlanResolutionSuite extends AnalysisTest {
...
-      }
-
-      assert(ae.getMessage
-        .contains(s"Unsupported partition transform: $transform"))
+      checkError(
+        exception = intercept[SparkUnsupportedOperationException] {
+          parseAndResolve(query)
+        },
+        errorClass = "_LEGACY_ERROR_TEMP_2067",
+        parameters = Map("transform" -> transform))
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:301]: image

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:301]: In fact, it throws S`parkUnsupportedOperationException`.

**Comment by panbingkun** [comment]
cc @MaxGekk

**Review by itholic — APPROVED** [review]
LGTM when CI pass

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -4312,5 +4312,10 @@
     "message" : [
       "Not enough memory to build and broadcast the table to all worker nodes. As a workaround, you can either disable broadcast by setting <autoBroadcastjoinThreshold> to -1 or increase the spark driver memory by setting <driverMemory> to a higher value<analyzeTblMsg>"
     ]
+  },
+  "_LEGACY_ERROR_TEMP_2251" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:4316]: Let's introduce a appropriate name as you are here.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done.

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -949,6 +949,11 @@
           "Literal for '<value>' of <type>."
         ]
       },
+      "MULTIPLE_BUCKET_TRANSFORM" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:952]: MULTIPLE_BUCKET_TRANSFORM -> MULTIPLE_BUCKET_TRANSFORMS

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -949,6 +949,11 @@
...
+        "message" : [
+          "TRANSFORM on multiple bucket."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:954]: Multiple bucket TRANSFORMs

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Pull request closed without merging**